protocols/v2/roles-logic-sv2/src/channel_logic/channel_factory.rs

```diff
@@ -0,0 +1,1075 @@
+            let success = OpenExtendedMiningChannelSuccess {
+                request_id,
+                channel_id,
+                target,
+                extranonce_size: max_extranonce_size,
+                extranonce_prefix,
+            };
+            self.extended_channels.insert(channel_id, success.clone());
+            let mut result = vec![Mining::OpenExtendedMiningChannelSuccess(success)];
+            if let Some((new_prev_hash, _)) = &self.last_prev_hash {
+                result.push(Mining::SetNewPrevHash(new_prev_hash.clone()))
+            };
+            if let Some((job, _)) = &self.last_valid_job {
+                result.push(Mining::NewExtendedMiningJob(job.clone()))
+            };
+            for (job, _) in &self.future_jobs {
+                result.push(Mining::NewExtendedMiningJob(job.clone()))
+            }
+            Some(result)
+        } else {
+            Some(vec![Mining::OpenMiningChannelError(
+                OpenMiningChannelError::new_max_target_out_of_range(request_id),
+            )])
+        }
+    }
+
+    pub fn on_new_prev_hash_from_tp(
+        &mut self,
+        m: &SetNewPrevHashFromTp<'static>,
+    ) -> Result<u32, Error> {
+        let job_id = self.job_creator.on_new_prev_hash(m).unwrap_or(0);
+        let new_prev_hash = SetNewPrevHash {
+            channel_id: 0,
+            job_id,
+            prev_hash: m.prev_hash.clone(),
+            min_ntime: m.header_timestamp,
+            nbits: m.n_bits,
+        };
+        self.on_new_prev_hash(new_prev_hash.clone())?;
+        Ok(job_id)
+    }
+
+    pub fn on_new_template(
+        &mut self,
+        m: &mut NewTemplate<'static>,
+    ) -> Result<HashMap<u32, Mining<'static>>, Error> {
+        let new_job = self.job_creator.on_new_template(m, true)?;
+        self.on_new_extended_mining_job(new_job)
+    }
+
+    pub fn on_submit_shares_extended(
+        &mut self,
+        m: SubmitSharesExtended<'static>,
+    ) -> Result<OnNewShare, Error> {
+        self.check_target(Share::Extended(m))
+    }
+    pub fn on_submit_shares_standard(
+        &mut self,
+        m: SubmitSharesStandard,
+        group_id: u32,
+    ) -> Result<OnNewShare, Error> {
+        self.check_target(Share::Standard((m, group_id)))
+    }
+
+    pub fn new_group_id(&mut self) -> u32 {
+        let new_id = self.ids.safe_lock(|ids| ids.new_group_id()).unwrap();
+        new_id
+    }
+    pub fn new_standard_id_for_hom(&mut self) -> u32 {
+        let hom_group_id = 0;
+        let new_id = self
+            .ids
+            .safe_lock(|ids| ids.new_channel_id(hom_group_id).unwrap())
+            .unwrap();
+        new_id
+    }
+
+    fn new_standard_channel_for_hom_downstream(
+        &mut self,
+        request_id: u32,
+        downstream_hash_rate: f32,
+        id: u32,
+    ) -> Result<Vec<Mining>, Error> {
+        let hom_group_id = 0;
+        let mut result = vec![];
+        // first unwrap: is ok to unwrap safe_lock results
```

Comment:
I don't see what unwraps these lines are referring to.

Reply:
ops I removed the unwraps but not the comments